### ISSUE TITLE
Upcasting the logits to fp32

### ISSUE THREAD

**Issue by cvejoski** [JetBrains-Research/opaque#862]
Reconsider the process of upcasting
